<a href="https://colab.research.google.com/github/vishnub0/AWS-WEP/blob/main/LR%2BRF_AWS15_Preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Imports

In [1]:
import pandas as pd
import numpy as np
!pip install seaborn
print("hello")

hello


#Data Preprocessing (make sure you upload the CSV files)

In [2]:
BALANCED_DATASET = 'diabetes_binary_5050split_health_indicators_BRFSS2015.csv'
df = pd.read_csv(BALANCED_DATASET)
df

Diabetes_binary  HighBP  HighChol  CholCheck   BMI  Smoker  Stroke  \
0                  0.0     1.0       0.0        1.0  26.0     0.0     0.0   
1                  0.0     1.0       1.0        1.0  26.0     1.0     1.0   
2                  0.0     0.0       0.0        1.0  26.0     0.0     0.0   
3                  0.0     1.0       1.0        1.0  28.0     1.0     0.0   
4                  0.0     0.0       0.0        1.0  29.0     1.0     0.0   
...                ...     ...       ...        ...   ...     ...     ...   
70687              1.0     0.0       1.0        1.0  37.0     0.0     0.0   
70688              1.0     0.0       1.0        1.0  29.0     1.0     0.0   
70689              1.0     1.0       1.0        1.0  25.0     0.0     0.0   
70690              1.0     1.0       1.0        1.0  18.0     0.0     0.0   
70691              1.0     1.0       1.0        1.0  25.0     0.0     0.0   

       HeartDiseaseorAttack  PhysActivity  Fruits  ...  AnyHealthcare  \
0                       0.0           1.0     0.0  ...            1.0   
1                       0.0           0.0     1.0  ...            1.0   
2                       0.0           1.0     1.0  ...            1.0   
3                       0.0           1.0     1.0  ...            1.0   
4                       0.0           1.0     1.0  ...            1.0   
...                     ...           ...     ...  ...            ...   
70687                   0.0           0.0     0.0  ...            1.0   
70688                   1.0           0.0     1.0  ...            1.0   
70689                   1.0           0.0     1.0  ...            1.0   
70690                   0.0           0.0     0.0  ...            1.0   
70691                   1.0           1.0     1.0  ...            1.0   

       NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex   Age  \
0              0.0      3.0       5.0      30.0       0.0  1.0   4.0   
1              0.0      3.0       0.0       0.0       0.0  1.0  12.0   
2              0.0      1.0       0.0      10.0       0.0  1.0  13.0   
3              0.0      3.0       0.0       3.0       0.0  1.0  11.0   
4              0.0      2.0       0.0       0.0       0.0  0.0   8.0   
...            ...      ...       ...       ...       ...  ...   ...   
70687          0.0      4.0       0.0       0.0       0.0  0.0   6.0   
70688          0.0      2.0       0.0       0.0       1.0  1.0  10.0   
70689          0.0      5.0      15.0       0.0       1.0  0.0  13.0   
70690          0.0      4.0       0.0       0.0       1.0  0.0  11.0   
70691          0.0      2.0       0.0       0.0       0.0  0.0   9.0   

       Education  Income  
0            6.0     8.0  
1            6.0     8.0  
2            6.0     8.0  
3            6.0     8.0  
4            5.0     8.0  
...          ...     ...  
70687        4.0     1.0  
70688        3.0     6.0  
70689        6.0     4.0  
70690        2.0     4.0  
70691        6.0     2.0  

[70692 rows x 22 columns]

Normalize columns

In [3]:
df_norm = df.copy()
features_to_normalize = ['BMI', 'GenHlth', 'MentHlth', 'PhysHlth', 'Age', 'Education', 'Income']
for feature in features_to_normalize:
  max = df_norm[feature].max()
  min = df_norm[feature].min()
  df_norm[feature] = (df_norm[feature] - min) / (max - min)

In [4]:
df_norm

Diabetes_binary  HighBP  HighChol  CholCheck       BMI  Smoker  Stroke  \
0                  0.0     1.0       0.0        1.0  0.162791     0.0     0.0   
1                  0.0     1.0       1.0        1.0  0.162791     1.0     1.0   
2                  0.0     0.0       0.0        1.0  0.162791     0.0     0.0   
3                  0.0     1.0       1.0        1.0  0.186047     1.0     0.0   
4                  0.0     0.0       0.0        1.0  0.197674     1.0     0.0   
...                ...     ...       ...        ...       ...     ...     ...   
70687              1.0     0.0       1.0        1.0  0.290698     0.0     0.0   
70688              1.0     0.0       1.0        1.0  0.197674     1.0     0.0   
70689              1.0     1.0       1.0        1.0  0.151163     0.0     0.0   
70690              1.0     1.0       1.0        1.0  0.069767     0.0     0.0   
70691              1.0     1.0       1.0        1.0  0.151163     0.0     0.0   

       HeartDiseaseorAttack  PhysActivity  Fruits  ...  AnyHealthcare  \
0                       0.0           1.0     0.0  ...            1.0   
1                       0.0           0.0     1.0  ...            1.0   
2                       0.0           1.0     1.0  ...            1.0   
3                       0.0           1.0     1.0  ...            1.0   
4                       0.0           1.0     1.0  ...            1.0   
...                     ...           ...     ...  ...            ...   
70687                   0.0           0.0     0.0  ...            1.0   
70688                   1.0           0.0     1.0  ...            1.0   
70689                   1.0           0.0     1.0  ...            1.0   
70690                   0.0           0.0     0.0  ...            1.0   
70691                   1.0           1.0     1.0  ...            1.0   

       NoDocbcCost  GenHlth  MentHlth  PhysHlth  DiffWalk  Sex       Age  \
0              0.0     0.50  0.166667  1.000000       0.0  1.0  0.250000   
1              0.0     0.50  0.000000  0.000000       0.0  1.0  0.916667   
2              0.0     0.00  0.000000  0.333333       0.0  1.0  1.000000   
3              0.0     0.50  0.000000  0.100000       0.0  1.0  0.833333   
4              0.0     0.25  0.000000  0.000000       0.0  0.0  0.583333   
...            ...      ...       ...       ...       ...  ...       ...   
70687          0.0     0.75  0.000000  0.000000       0.0  0.0  0.416667   
70688          0.0     0.25  0.000000  0.000000       1.0  1.0  0.750000   
70689          0.0     1.00  0.500000  0.000000       1.0  0.0  1.000000   
70690          0.0     0.75  0.000000  0.000000       1.0  0.0  0.833333   
70691          0.0     0.25  0.000000  0.000000       0.0  0.0  0.666667   

       Education    Income  
0            1.0  1.000000  
1            1.0  1.000000  
2            1.0  1.000000  
3            1.0  1.000000  
4            0.8  1.000000  
...          ...       ...  
70687        0.6  0.000000  
70688        0.4  0.714286  
70689        1.0  0.428571  
70690        0.2  0.428571  
70691        1.0  0.142857  

[70692 rows x 22 columns]

#Data Visualization

In [5]:
"""
# Visualizations

import seaborn as sns
import matplotlib.pyplot as plt

# Show the first few rows of the dataset
print(df.head())

# Statistical summary

# If you are using a pandas dataframe (pd.DataFrame), or just "df" as it will appear
# in the coming notebooks, you can also use df.describe() to achieve the same thing as what the line below does.
# The line below in fact is very specific to this toy dataset.
print(df.describe())
"""

'\n# Visualizations\n\nimport seaborn as sns\nimport matplotlib.pyplot as plt\n\n# Show the first few rows of the dataset\nprint(df.head())\n\n# Statistical summary\n\n# If you are using a pandas dataframe (pd.DataFrame), or just "df" as it will appear\n# in the coming notebooks, you can also use df.describe() to achieve the same thing as what the line below does.\n# The line below in fact is very specific to this toy dataset.\nprint(df.describe())\n'

In [6]:
"""
# Pairplot to visualize the relationships between features
df.columns = df.columns.str.strip()
print(df.columns.tolist())
cols_to_plot = ['HighBP', 'HighChol', 'BMI', 'PhysActivity', 'Age', 'Income', 'Diabetes_binary']

# Sample 2000 rows for quicker plotting
df_sampled = df[cols_to_plot].sample(n=2000, random_state=42)
sns.pairplot(df_sampled, hue='Diabetes_binary', palette='coolwarm')
plt.suptitle('Pairplot (Sampled Data)', y=1.02)
plt.show()
"""

"\n# Pairplot to visualize the relationships between features\ndf.columns = df.columns.str.strip()\nprint(df.columns.tolist())\ncols_to_plot = ['HighBP', 'HighChol', 'BMI', 'PhysActivity', 'Age', 'Income', 'Diabetes_binary']\n\n# Sample 2000 rows for quicker plotting\ndf_sampled = df[cols_to_plot].sample(n=2000, random_state=42)\nsns.pairplot(df_sampled, hue='Diabetes_binary', palette='coolwarm')\nplt.suptitle('Pairplot (Sampled Data)', y=1.02)\nplt.show()\n"

In [7]:
"""
# Box Plot to Compare BMI across diabetes status
plt.figure(figsize=(8, 6))
sns.boxplot(x='Diabetes_binary', y='BMI', data=df)
plt.title('Box Plot of BMI by Diabetes Status')
plt.xlabel('Diabetes Status (0 = No, 1 = Yes)')
plt.ylabel('BMI')
plt.show()
"""

"\n# Box Plot to Compare BMI across diabetes status\nplt.figure(figsize=(8, 6))\nsns.boxplot(x='Diabetes_binary', y='BMI', data=df)\nplt.title('Box Plot of BMI by Diabetes Status')\nplt.xlabel('Diabetes Status (0 = No, 1 = Yes)')\nplt.ylabel('BMI')\nplt.show()\n"

In [8]:
"""
# Histogram of BMI distribution
plt.figure(figsize=(8, 6))
sns.histplot(df['BMI'], kde=True, bins=30)
plt.title('Distribution of BMI')
plt.xlabel('BMI')
plt.ylabel('Frequency')
plt.show()
"""

"\n# Histogram of BMI distribution\nplt.figure(figsize=(8, 6))\nsns.histplot(df['BMI'], kde=True, bins=30)\nplt.title('Distribution of BMI')\nplt.xlabel('BMI')\nplt.ylabel('Frequency')\nplt.show()\n"

In [9]:
"""
# Heatmap of Correlation Matrix
plt.figure(figsize=(14, 12))
corr_matrix = df.corr()
sns.heatmap(corr_matrix, annot=True, cmap='coolwarm', fmt=".2f", linewidths=0.5)
plt.title('Correlation Heatmap of Health Indicators')
plt.show()
"""

'\n# Heatmap of Correlation Matrix\nplt.figure(figsize=(14, 12))\ncorr_matrix = df.corr()\nsns.heatmap(corr_matrix, annot=True, cmap=\'coolwarm\', fmt=".2f", linewidths=0.5)\nplt.title(\'Correlation Heatmap of Health Indicators\')\nplt.show()\n'

#Feature tuning (removing columns)

In [10]:
cols_to_remove = ['Smoker', 'Fruits', 'Veggies', 'HvyAlcoholConsump', 'AnyHealthcare', 'NoDocbcCost',
                  'MentHlth', 'Sex', 'Stroke', 'PhysActivity', 'Income', 'Education']
df_norm = df_norm.drop(cols_to_remove, axis='columns')
df_norm

Diabetes_binary  HighBP  HighChol  CholCheck       BMI  \
0                  0.0     1.0       0.0        1.0  0.162791   
1                  0.0     1.0       1.0        1.0  0.162791   
2                  0.0     0.0       0.0        1.0  0.162791   
3                  0.0     1.0       1.0        1.0  0.186047   
4                  0.0     0.0       0.0        1.0  0.197674   
...                ...     ...       ...        ...       ...   
70687              1.0     0.0       1.0        1.0  0.290698   
70688              1.0     0.0       1.0        1.0  0.197674   
70689              1.0     1.0       1.0        1.0  0.151163   
70690              1.0     1.0       1.0        1.0  0.069767   
70691              1.0     1.0       1.0        1.0  0.151163   

       HeartDiseaseorAttack  GenHlth  PhysHlth  DiffWalk       Age  
0                       0.0     0.50  1.000000       0.0  0.250000  
1                       0.0     0.50  0.000000       0.0  0.916667  
2                       0.0     0.00  0.333333       0.0  1.000000  
3                       0.0     0.50  0.100000       0.0  0.833333  
4                       0.0     0.25  0.000000       0.0  0.583333  
...                     ...      ...       ...       ...       ...  
70687                   0.0     0.75  0.000000       0.0  0.416667  
70688                   1.0     0.25  0.000000       1.0  0.750000  
70689                   1.0     1.00  0.000000       1.0  1.000000  
70690                   0.0     0.75  0.000000       1.0  0.833333  
70691                   1.0     0.25  0.000000       0.0  0.666667  

[70692 rows x 10 columns]

#Splitting data

In [11]:
X = df_norm.drop('Diabetes_binary', axis='columns')
y = df_norm['Diabetes_binary']

In [12]:
X

HighBP  HighChol  CholCheck       BMI  HeartDiseaseorAttack  GenHlth  \
0         1.0       0.0        1.0  0.162791                   0.0     0.50   
1         1.0       1.0        1.0  0.162791                   0.0     0.50   
2         0.0       0.0        1.0  0.162791                   0.0     0.00   
3         1.0       1.0        1.0  0.186047                   0.0     0.50   
4         0.0       0.0        1.0  0.197674                   0.0     0.25   
...       ...       ...        ...       ...                   ...      ...   
70687     0.0       1.0        1.0  0.290698                   0.0     0.75   
70688     0.0       1.0        1.0  0.197674                   1.0     0.25   
70689     1.0       1.0        1.0  0.151163                   1.0     1.00   
70690     1.0       1.0        1.0  0.069767                   0.0     0.75   
70691     1.0       1.0        1.0  0.151163                   1.0     0.25   

       PhysHlth  DiffWalk       Age  
0      1.000000       0.0  0.250000  
1      0.000000       0.0  0.916667  
2      0.333333       0.0  1.000000  
3      0.100000       0.0  0.833333  
4      0.000000       0.0  0.583333  
...         ...       ...       ...  
70687  0.000000       0.0  0.416667  
70688  0.000000       1.0  0.750000  
70689  0.000000       1.0  1.000000  
70690  0.000000       1.0  0.833333  
70691  0.000000       0.0  0.666667  

[70692 rows x 9 columns]

In [13]:
y

0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
70687    1.0
70688    1.0
70689    1.0
70690    1.0
70691    1.0
Name: Diabetes_binary, Length: 70692, dtype: float64

#Implementing Logistic Regression and Random Forest

In [14]:
!pip install imbalanced-learn
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report
from imblearn.over_sampling import SMOTE

In [15]:
def run_test_0(X, y):
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    lr = LogisticRegression(max_iter=1000)
    lr.fit(X_train, y_train)
    lr_preds = lr.predict(X_test)
    print("Test 0 - Logistic Regression (Baseline):")
    print("Test Accuracy:", accuracy_score(y_test, lr_preds))
    print(classification_report(y_test, lr_preds))

    rf = RandomForestClassifier(random_state=42)
    rf.fit(X_train, y_train)
    rf_preds = rf.predict(X_test)
    print("\nTest 0 - Random Forest (Baseline):")
    print("Test Accuracy:", accuracy_score(y_test, rf_preds))
    print(classification_report(y_test, rf_preds))

In [16]:
def run_test_1(X, y):
    X = X.copy()
    X['BMI_Age'] = X['BMI'] * X['Age']
    X['BMI2'] = X['BMI'] ** 2
    X['Age2'] = X['Age'] ** 2

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    X_train, _, y_train, _ = train_test_split(X_train, y_train, test_size=0.25, random_state=42)

    sm = SMOTE(random_state=42)
    X_train_res, y_train_res = sm.fit_resample(X_train, y_train)

    lr = LogisticRegression(penalty='l1', solver='liblinear', class_weight='balanced', max_iter=1000)
    lr.fit(X_train_res, y_train_res)
    print("Test 1 - Logistic Regression (L1):")
    print("Test Accuracy:", accuracy_score(y_test, lr.predict(X_test)))
    print(classification_report(y_test, lr.predict(X_test)))

    param_grid = {
        'n_estimators': [100],
        'max_depth': [10, 20],
        'min_samples_split': [2, 5],
        'min_samples_leaf': [1, 2],
        'class_weight': ['balanced']
    }
    rf_search = GridSearchCV(RandomForestClassifier(random_state=42), param_grid, cv=3, n_jobs=1)
    rf_search.fit(X_train_res, y_train_res)
    print("\nTest 1 - Random Forest:")
    print("Best Params:", rf_search.best_params_)
    print("Test Accuracy:", accuracy_score(y_test, rf_search.predict(X_test)))
    print(classification_report(y_test, rf_search.predict(X_test)))

In [17]:
def run_test_2(X, y):
    from sklearn.model_selection import train_test_split, GridSearchCV
    from sklearn.linear_model import LogisticRegression
    from sklearn.ensemble import RandomForestClassifier
    from sklearn.metrics import accuracy_score, classification_report
    from imblearn.over_sampling import SMOTE

    X_train_base, X_test_base, y_train_base, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    X_train_base, _, y_train_base, _ = train_test_split(X_train_base, y_train_base, test_size=0.25, random_state=42)

    def add_features(df):
        df = df.copy()
        df['BMI_Age'] = df['BMI'] * df['Age']
        df['BMI2'] = df['BMI'] ** 2
        df['Age2'] = df['Age'] ** 2
        df['GenHlth_PhysHlth'] = df['GenHlth'] * df['PhysHlth']
        df['Heart_BP'] = df['HeartDiseaseorAttack'] * df['HighBP']
        return df

    X_train_lr = add_features(X_train_base)
    X_test_lr = add_features(X_test_base)

    sm = SMOTE(random_state=42)
    X_train_lr_res, y_train_lr_res = sm.fit_resample(X_train_lr, y_train_base)
    X_train_rf_res, y_train_rf_res = sm.fit_resample(X_train_base, y_train_base)

    lr = LogisticRegression(penalty='l2', solver='lbfgs', C=1.0, class_weight='balanced', max_iter=1000)
    lr.fit(X_train_lr_res, y_train_lr_res)
    print("Test 2 - Logistic Regression (L2):")
    print("Test Accuracy:", accuracy_score(y_test, lr.predict(X_test_lr)))
    print(classification_report(y_test, lr.predict(X_test_lr)))

    param_grid = {
        'n_estimators': [100],
        'max_depth': [10, 15],
        'min_samples_split': [2],
        'min_samples_leaf': [1],
        'class_weight': ['balanced']
    }

    rf_search = GridSearchCV(RandomForestClassifier(random_state=42), param_grid, cv=3, n_jobs=1)
    rf_search.fit(X_train_rf_res, y_train_rf_res)
    print("\nTest 2 - Random Forest:")
    print("Best Params:", rf_search.best_params_)
    print("Test Accuracy:", accuracy_score(y_test, rf_search.predict(X_test_base)))
    print(classification_report(y_test, rf_search.predict(X_test_base)))

In [18]:
def run_test_3(X, y):
    from sklearn.model_selection import train_test_split, GridSearchCV
    from sklearn.linear_model import LogisticRegression
    from sklearn.ensemble import RandomForestClassifier
    from sklearn.metrics import accuracy_score, classification_report
    from imblearn.over_sampling import SMOTE

    X = X.copy()
    X['BMI_Age'] = X['BMI'] * X['Age']
    X['HealthScore'] = X['GenHlth'] + X['PhysHlth']

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    X_train, _, y_train, _ = train_test_split(X_train, y_train, test_size=0.25, random_state=42)

    sm = SMOTE(random_state=42)
    X_train_res, y_train_res = sm.fit_resample(X_train, y_train)

    lr = LogisticRegression(penalty='l2', solver='lbfgs', C=1.2, class_weight='balanced', max_iter=1000)
    lr.fit(X_train_res, y_train_res)
    lr_preds = lr.predict(X_test)

    print("Test 3 - Logistic Regression (L2, minimal features):")
    print("Test Accuracy:", accuracy_score(y_test, lr_preds))
    print(classification_report(y_test, lr_preds))

    param_grid = {
        'n_estimators': [100, 200],
        'max_depth': [10, 15],
        'min_samples_split': [2, 5],
        'min_samples_leaf': [1],
        'class_weight': ['balanced']
    }

    rf_search = GridSearchCV(RandomForestClassifier(random_state=42), param_grid, cv=3, n_jobs=1)
    rf_search.fit(X_train, y_train)

    best_rf = rf_search.best_estimator_
    rf_preds = best_rf.predict(X_test)

    print("\nTest 3 - Optimized Random Forest:")
    print("Best Params:", rf_search.best_params_)
    print("Test Accuracy:", accuracy_score(y_test, rf_preds))
    print(classification_report(y_test, rf_preds))

In [19]:
def run_test_4(X, y):
    from sklearn.model_selection import train_test_split, GridSearchCV
    from sklearn.linear_model import LogisticRegression
    from sklearn.ensemble import RandomForestClassifier
    from sklearn.metrics import accuracy_score, classification_report
    from imblearn.over_sampling import SMOTE

    X = X.copy()
    X['BMI_Age'] = X['BMI'] * X['Age']
    X['HealthScore'] = X['GenHlth'] + X['PhysHlth']

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    X_train, _, y_train, _ = train_test_split(X_train, y_train, test_size=0.25, random_state=42)

    sm = SMOTE(random_state=42)
    X_train_res, y_train_res = sm.fit_resample(X_train, y_train)

    lr = LogisticRegression(
        penalty='elasticnet',
        solver='saga',
        l1_ratio=0.2,  # 20% L1, 80% L2
        C=1.2,
        class_weight='balanced',
        max_iter=2000
    )
    lr.fit(X_train_res, y_train_res)
    lr_preds = lr.predict(X_test)

    print("Test 4 - Logistic Regression (ElasticNet):")
    print("Test Accuracy:", accuracy_score(y_test, lr_preds))
    print(classification_report(y_test, lr_preds))

    param_grid = {
        'n_estimators': [100, 200],
        'max_depth': [10, 15],
        'min_samples_split': [2],
        'min_samples_leaf': [1, 2],
        'max_features': ['sqrt'],  # NEW
        'class_weight': ['balanced']
    }

    rf_search = GridSearchCV(RandomForestClassifier(random_state=42), param_grid, cv=3, n_jobs=1)
    rf_search.fit(X_train, y_train)

    best_rf = rf_search.best_estimator_
    rf_preds = best_rf.predict(X_test)

    print("\nTest 4 - Optimized Random Forest:")
    print("Best Params:", rf_search.best_params_)
    print("Test Accuracy:", accuracy_score(y_test, rf_preds))
    print(classification_report(y_test, rf_preds))

In [20]:
run_test_0(X, y)
print("\n" + "="*70 + "\n")

run_test_1(X, y)
print("\n" + "="*70 + "\n")

run_test_2(X, y)
print("\n" + "="*70 + "\n")

run_test_3(X, y)
print("\n" + "="*70 + "\n")

run_test_4(X, y)

Test 0 - Logistic Regression (Baseline):
Test Accuracy: 0.7455972841077869
              precision    recall  f1-score   support

         0.0       0.76      0.72      0.74      7090
         1.0       0.73      0.77      0.75      7049

    accuracy                           0.75     14139
   macro avg       0.75      0.75      0.75     14139
weighted avg       0.75      0.75      0.75     14139


Test 0 - Random Forest (Baseline):
Test Accuracy: 0.7079708607397978
              precision    recall  f1-score   support

         0.0       0.72      0.69      0.70      7090
         1.0       0.70      0.72      0.71      7049

    accuracy                           0.71     14139
   macro avg       0.71      0.71      0.71     14139
weighted avg       0.71      0.71      0.71     14139



Test 1 - Logistic Regression (L1):
Test Accuracy: 0.7486385175755004
              precision    recall  f1-score   support

         0.0       0.76      0.73      0.74      7090
         1.0       0.

**Logistic Regression – Test Insights**

Test 1:
-Regularization: L1 (Lasso)
-Feature Engineering: BMI*Age, BMI², Age²
-Accuracy: 0.7486
-Insight: L1 promotes sparsity by zeroing out weak features. This helped prevent overfitting but may have removed useful predictors.

Test 2:
-Regularization: L2 (Ridge)
-Feature Engineering: BMI*Age, BMI², Age², GenHlth*PhysHlth, HeartDisease*BP
-Accuracy: 0.7489 ✅ (Best)
-Insight: L2 keeps all features and distributes weights evenly. Added health interactions improved performance and generalization.

Test 3:
-Regularization: L2
-Feature Engineering: BMI*Age, GenHlth + PhysHlth
-Accuracy: 0.7471
-Insight: Fewer features slightly hurt accuracy, showing LR benefits from rich feature interactions.

Test 4:
-Regularization: ElasticNet (80% L2, 20% L1)
-Feature Engineering: Same as Test 1
-Accuracy: 0.7473
-Insight: Slightly underperformed L2. The L1 part may have trimmed helpful variables.

**Logistic Summary:**
Test 2 is best. Use L2 regularization and rich interaction features for strongest logistic regression results.

**Random Forest – Test Insights**

Test 1:
-Params: 100 trees, max_depth=10, baseline SMOTE
-Accuracy: 0.7449
-Insight: Solid baseline, handled imbalance well with class_weight.

Test 2:
-Params: Same as Test 1, different SMOTE sample
-Accuracy: 0.7467
-Insight: Slight improvement. RF is sensitive to how SMOTE modifies training distribution.

Test 3:
-Params: 200 trees (higher n_estimators)
-Accuracy: 0.7471 ✅ (Best)
-Insight: More trees stabilized predictions and slightly improved accuracy.

Test 4:
-Params: 200 trees, max_features='sqrt', min_samples_leaf=2
-Accuracy: 0.7458
-Insight: Added regularization slightly reduced flexibility, hurting performance.

**Random Forest Summary:**
Test 3 is best. More trees improved stability. Over-regularizing (Test 4) didn’t help.

**LOGISTIC REGRESSION**

Test 1 — L1 Regularization
-Change: penalty='l1', solver='liblinear'
-Why: L1 (Lasso) adds sparsity by zeroing out less important features. It's good for feature selection and reducing overfitting.
-Feature Engineering:
-BMI * Age: Interaction term to capture compounding risk.
-BMI^2, Age^2: Capture nonlinear effects.
-Impact: Strong generalization (74.86%) but possibly discarded some informative features due to L1's aggressive sparsity.

Test 2 — L2 Regularization + Advanced Features
-Change: penalty='l2', solver='lbfgs', C=1.0
-Why: L2 (Ridge) doesn't zero out features, keeping more information. Better for stable models.
-Feature Engineering (expanded):
-Adds GenHlth * PhysHlth: interaction between general and physical health.
-Adds HeartDisease * HighBP: captures compound cardiovascular risk.
-Impact: Best accuracy (74.89%). Rich feature set improved model understanding of health correlations.

Test 3 — L2 Regularization + Minimal Features
-Change: Kept L2, but reduced features to only:
-BMI * Age
-GenHlth + PhysHlth (simple additive)
-Why: To test whether simpler features would suffice or improve generalization.
-Impact: Accuracy dropped slightly (74.71%). Indicates richer engineered features from Test 2 were more effective.

Test 4 — ElasticNet Regularization
-Change: penalty='elasticnet', solver='saga', l1_ratio=0.2 (20% L1, 80% L2)
-Why: Hybrid approach to balance between L1's feature selection and L2's stability.
-Feature Engineering: Same as Test 1 (BMI, Age polynomials).
-Impact: Accuracy (74.73%) was slightly below L2. Likely because partial L1 still pruned useful features.

**RANDOM FOREST**

Test 1 — Baseline Random Forest
-Change:
-n_estimators=100
-max_depth=10
-class_weight='balanced'
-Why: Basic setup to ensure trees don’t grow too deep and overfit. Balanced weights handle class imbalance.
-Impact: Strong baseline (74.49%), recall favored diabetic class (1).

Test 2 — Same RF, Different SMOTE Data
-Change: Same hyperparameters as Test 1.
-Why: Purposefully used different resampled data to test consistency.
-Impact: Slight bump in accuracy (74.67%), shows RF is somewhat sensitive to training distribution from SMOTE.

Test 3 — Increased Estimators
-Change: n_estimators=200 (up from 100)
-Why: More trees reduce variance and stabilize predictions.
-Impact: Best RF performance (74.71%). Indicates more trees led to more reliable generalization.

Test 4 — Added Regularization
-Change:
-min_samples_leaf=2: Avoid overly specific splits
-max_features='sqrt': Random subset of features per split (common trick)
-Why: Intended to reduce overfitting by controlling tree complexity.
-Impact: Slightly lower accuracy (74.58%). Likely over-regularized; constrained trees couldn't capture enough nuance.

**Logistic Regression Progression**

-Test 1 – L1 Regularization (Baseline)
Why Start Here: L1 is great for initial feature filtering. It helps find which features are most important by zeroing out weaker ones.
-Goal: Establish a simple, regularized model that fights overfitting and gives a decent baseline.
-Result: ~74.86% accuracy. Solid start, but L1 can discard useful signals.

Test 2 – L2 Regularization + Advanced Features
-Why: L1 might’ve been too aggressive. So, switch to L2 which keeps all features and penalizes large weights without deleting them.
-What Changed:
-Regularization switched to L2 (penalty='l2').
-Added new engineered features: GenHlth * PhysHlth, HeartDisease * HighBP, etc.
-Goal: Give the model more expressive power and retain more feature interactions.
-Result: Highest logistic regression accuracy (~74.89%). Feature richness helped.

Test 3 – L2, Minimal Features
-Why: See if fewer features would generalize better. Often simpler models avoid overfitting.
-What Changed:
-Used only BMI_Age and GenHlth + PhysHlth
-Goal: Reduce model complexity, test if rich features were necessary.
-Result: Accuracy dipped (~74.71%). Conclusion: extra features from Test 2 helped.

Test 4 – ElasticNet (L1 + L2 Hybrid)
-Why: Combine benefits of both L1 and L2.
-What Changed:
-Used penalty='elasticnet' with l1_ratio=0.2
-Goal: Retain most features while still trimming noise.
-Result: Slight improvement over minimal L2, but worse than full L2 (~74.73%). Suggests L1 still pruned useful info.

**Random Forest Progression**

-Test 1 – Baseline RF
-Why Start Here: Use shallow trees with basic grid (max_depth=10, n_estimators=100)
-Goal: Prevent overfitting, especially on small splits. Use class balancing to deal with imbalance.
-Result: ~74.49%. Good base.

Test 2 – Same RF, Different SMOTE Split
-Why: Explore if changes in the SMOTE resample affect performance.
-What Changed: Exact same model params, but new random sample from SMOTE.
-Result: Slight boost (~74.67%). Confirmed SMOTE variation matters.

Test 3 – More Estimators
-Why: Boost accuracy by increasing number of trees (n_estimators=200)
-Goal: Reduce variance, get more stable performance.
-Result: Best RF performance (~74.71%). More trees helped generalization.

Test 4 – Add Regularization to Trees
-Why: Avoid overfitting via:
-min_samples_leaf=2 (forces broader splits)
-max_features='sqrt' (use subset of features per split)
-Goal: Prevent overly specific trees from capturing noise.
-Result: Slight drop (~74.58%). Suggests model was over-regularized and lost nuance.